### PR TITLE
Add RTL_CLIMB_MIN test for ArduPlane

### PR DESCRIPTION
**Pull request by peterbarker**
This clearly reproduces the flight characteristics in https://github.com/ArduPilot/ardupilot/issues/16532

image

The test *fails* as it is expected the vehicle should maintain the higher altitude.  The code is quite clear clear - we only add the extra altitude until we've finished the climb:
```
        if (control_mode == &mode_rtl && !rtl.done_climb && g2.rtl_climb_min > 0) {
            // ensure we do the initial climb in RTL. We add an extra
            // 10m in the demanded height to push TECS to climb
            // quickly
            target_alt = MAX(target_alt, prev_WP_loc.alt + (g2.rtl_climb_min+10)*100);
        }
```

### CODE REVIEW

**Comment by WickedShell** [comment]
Personally I'd rather have this be a bit more copter like where it was actually a `RTL_MIN_ALTITUDE` where it will climb to that height above home at a minimum and do so quickly with the roll limit, or climb to the RTL altitude whichever was higher and hold it.

Also if we are asking for scope creep I'd *very strongly* prefer that if we have a climb commanded by RTL we do it at max climb rate rather then build a slope to the rally point like we currently do. It makes planning around tall obstructions difficult.

**Comment by peterbarker** [comment]
On Mon, 8 Feb 2021, WickedShell wrote:

> Also if we are asking for scope creep I'd very strongly prefer that if we have a climb commanded by RTL we do it at max climb rate rather then build a

There's no scope creep here at all.  All I've done is replicated the 
flight behaviour, not attempted to fix it :-P

I'm not even sure that the current behaviour isn't what the original 
author envisaged!

**Comment by skorokithakis** [comment]
I *think* this is a bug in the original code. I believe the original author meant to only add the 10m during the initial climb so that the TECS climbs faster, but then didn't realize that the new target altitude needed to account for the extra addition after the climb was done.

**Comment by peterbarker** [comment]
I've added a change to the ArduPlane code in this PR now.  Now the flight profile looks like this:

image

So after the climb is done we reset the previous wp location to our current location and recalculate our glide slope home.

i.e. we do NOT maintain our new altitude all the way home; we descend linearly to our parameterised RTL altitude.

File: Tools/autotest/arduplane.py [hunk 1]
```diff
@@ -1427,6 +1427,34 @@ def deadreckoning(self):
...
+        self.wait_ready_to_arm()
+        rtl_climb_min = 100
+        self.set_parameter("RTL_CLIMB_MIN", rtl_climb_min)
+        takeoff_alt = 50
+        self.takeoff(takeoff_alt)
+        self.change_mode('CRUISE')
+        self.wait_distance_to_home(1000, 1500, timeout=60)
+        self.change_mode('RTL')
+        self.wait_altitude(90, 110, relative=True)
+        expected_alt = self.get_parameter("ALT_HOLD_RTL")/100.0
```
Comment (tridge) [inline Tools/autotest/arduplane.py:1440]: maybe check half way if it is within a margin of mid-point of RTL alt and climb min + start alt

**Review by tridge — CHANGES_REQUESTED** [review]

**Comment by skorokithakis** [comment]
Thanks, that's what I would expect. What's the reason for the bump in the graph, instead of completely linear descent?

**Comment by peterbarker** [comment]
> Thanks, that's what I would expect. What's the reason for the bump in the graph, instead of completely linear descent?

Poor TECS tuning on the simulated aircraft is our guess.

**Comment by skorokithakis** [comment]
Is there anything else that needs to be done here? Can I help?

**Comment by peterbarker** [comment]
@skorokithakis tridge just asked for the test to be augmented, which I've now done.  

I'm going to merge this - but if you can test fly it and report it fixes the problem, please do :-)

**Comment by skorokithakis** [comment]
Thank you, I can! Is master in a stable state/safe to fly, generally?

**Comment by peterbarker** [comment]
On Sat, 13 Feb 2021, Stavros Korokithakis wrote:

> Thank you, I can! Is master in a stable state/safe to fly, generally?

Generally, if you're a confident manual pilot, yes.  Just remember your 
"two mistakes high" :-)  (we don't generally tell people to fly master, 
but its generally fine)

**Comment by skorokithakis** [comment]
Ah good, I do usually fly manual, so it should be okay. I just won't stray out of video range, thanks!

**Pull request merged**